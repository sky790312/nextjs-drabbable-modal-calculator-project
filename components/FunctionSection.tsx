import React from 'react'
import styled from 'styled-components/'
import { FlexCenterContainer } from '@/GlobalStyles'

export const FunctionSection: React.FC = () => {
  return (
    <FunctionSectionContainer>
      <FlexCenterContainer>
        <Button>AC</Button>
      </FlexCenterContainer>
      <FlexCenterContainer>
        <Button>+/-</Button>
      </FlexCenterContainer>
      <FlexCenterContainer>
        <Button>%</Button>
      </FlexCenterContainer>
    </FunctionSectionContainer>

  )
}

const FunctionSectionContainer = styled.div`
  display: flex;
  flex-wrap: wrap;

  > div {
    width: 33.33%;
    padding: 10px;

    > button {
      color: ${({ theme }) => theme.colors.black};
      background-color: ${({ theme }) => theme.colors.gray};
      font-size: 24px;
    }
  }
`

const Button = styled.button`
  border-radius: 50%;
  border: none;
  width: 80px;
  height: 80px;

  &.x2 {
    width: 80%;
    border-radius: 40px;
  }
`