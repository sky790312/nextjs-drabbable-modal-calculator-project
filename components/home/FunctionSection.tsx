import React from 'react'
import styled from 'styled-components/'
import { FlexCenterContainer } from '@/GlobalStyles'
import { Dispatch } from 'redux'
import { calculatorActions, initialState } from '@/store/state.calculator'
import { useSelector, shallowEqual, useDispatch } from 'react-redux'
import { RootState } from '@/store/index'
import { Button } from '@/components/uiComponents/Button'
import { NUM_FUNCTIONS } from '@/constants'

export const FunctionSection: React.FC = React.memo(() => {
  const dispatch: Dispatch = useDispatch()
  const { reset, setDisplayValue } = calculatorActions
  const displayValue = useSelector(
    (state: RootState) => state.calculator.displayValue,
    shallowEqual,
  )

  const numFunctionHandler = {
    AC: () => handleReset(),
    '+/-': () => handleToggleSign(),
    '%': () => handlePercent(),
  }

  const handleReset = () => {
    dispatch(reset())
  }

  const handleToggleSign = () => {
    const newDisplayValue = (parseFloat(displayValue) * -1).toString()
    dispatch(setDisplayValue(newDisplayValue))
  }

  const handlePercent = () => {
    if (displayValue === initialState.displayValue) {
      return
    }

    const fixedDigits = displayValue.replace(/^-?\d*\.?/, '')
    const newValue = parseFloat(displayValue) / 100
    const newDisplayValue = String(newValue.toFixed(fixedDigits.length + 2))
    dispatch(setDisplayValue(newDisplayValue))
  }

  return (
    <FunctionSectionContainer>
      {NUM_FUNCTIONS.map(numFunction => (
        <FlexCenterContainer key={numFunction}>
          <Button onClick={() => numFunctionHandler[numFunction]()}>
            {numFunction}
          </Button>
        </FlexCenterContainer>
      ))}
    </FunctionSectionContainer>
  )
})

const FunctionSectionContainer = styled.div`
  display: flex;
  flex-wrap: wrap;

  > div {
    width: 33.33%;
    padding: 10px;
  }
`
