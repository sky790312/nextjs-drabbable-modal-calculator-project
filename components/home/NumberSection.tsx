import React from 'react'
import styled from 'styled-components/'
import { FlexCenterContainer } from '@/GlobalStyles'
import { Dispatch } from 'redux'
import { calculatorActions, initialState } from '@/store/state.calculator'
import { useSelector, shallowEqual, useDispatch } from 'react-redux'
import { RootState } from '@/store/index'
import { Button } from '@/components/uiComponents/Button'
import { NUMBERS } from '@/constants'

export const NumberSection: React.FC = React.memo(() => {
  const displayValue = useSelector(
    (state: RootState) => state.calculator.displayValue,
    shallowEqual,
  )
  const operator = useSelector(
    (state: RootState) => state.calculator.operator,
    shallowEqual,
  )

  const dispatch: Dispatch = useDispatch()
  const { setOperator, setDisplayValue } = calculatorActions

  const numberHandler = {
    num: (num: number) => handleNumber(num),
    dot: () => handleDot(),
  }

  const handleNumber = (num: number) => {
    const numString = num.toString()
    if (operator) {
      dispatch(setDisplayValue(numString))
      dispatch(setOperator(''))
    } else {
      const newDisplayValue =
        displayValue === initialState.displayValue
          ? numString
          : `${displayValue}${numString}`
      dispatch(setDisplayValue(newDisplayValue))
    }
  }

  const handleDot = () => {
    if (/\./.test(displayValue)) {
      return
    }

    dispatch(setDisplayValue(`${displayValue}.`))
  }

  return (
    <NumberSectionContainer>
      {NUMBERS.map(num => (
        <FlexCenterContainer key={num} className={!num && 'x2'}>
          <Button
            buttonStyle={'darkGray'}
            onClick={() =>
              isNaN(Number(num))
                ? numberHandler['dot']()
                : numberHandler['num'](Number(num))
            }
            size={!num ? 'block' : 'md'}
          >
            {num}
          </Button>
        </FlexCenterContainer>
      ))}
    </NumberSectionContainer>
  )
})

const NumberSectionContainer = styled.div`
  display: flex;
  flex-wrap: wrap-reverse;

  > div {
    width: 33.33%;
    padding: 10px;

    &.x2 {
      width: 66.67%;
    }
  }
`
