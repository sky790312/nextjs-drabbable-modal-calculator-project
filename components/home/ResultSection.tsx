import React from 'react'
import styled from 'styled-components/'
import { useSelector, shallowEqual } from 'react-redux'
import { RootState } from '@/store/index'
import { addCommas } from '@/utils'

export const ResultSection: React.FC = () => {
  const displayValue = useSelector(
    (state: RootState) => state.calculator.displayValue,
    shallowEqual,
  )

  return (
    <ResultSectionContainer>
      {addCommas(displayValue)}
    </ResultSectionContainer>
  )
}

const ResultSectionContainer = styled.div`
  height: 150px;
  text-align: right;
  color: ${({ theme }) => theme.colors.white};
  padding: 60px 20px 0;
  font-size: 64px;
  overflow: auto;
`
