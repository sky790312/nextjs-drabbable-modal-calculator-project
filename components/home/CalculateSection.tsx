import React, { useContext } from 'react'
import styled from 'styled-components/'
import { FunctionSection } from '@/components/home/FunctionSection'
import { NumberSection } from '@/components/home/NumberSection'
import { OperationSection } from '@/components/home/OperationSection'
import { ResultSection } from '@/components/home/ResultSection'
import { AppContext } from '@/components/appCommon/AppContext'

export const CalculateSection: React.FC = () => {
  const context = useContext(AppContext)
  const { deviceInfo } = context

  return (
    <CalculateSectionContainer deviceInfo={deviceInfo}>
      <ResultSection />
      <ManipulateSection>
        <div>
          <FunctionSection />
          <NumberSection />
        </div>
        <OperationSection />
      </ManipulateSection>
    </CalculateSectionContainer>
  )
}

const CalculateSectionContainer = styled.div<{ deviceInfo: 'pc' | 'mobile' }>`
  overflow: auto;
  background-image: linear-gradient(
    ${({ theme }) => theme.colors.lightBlue},
    ${({ theme }) => theme.colors.darkBlue}
  );
  height: 650px;
  border-radius: 5px;

  ${props => {
    if (props.deviceInfo === 'pc') {
      return `
        width: 360px;
        transition: 0.3s;

        @media (min-width: 1140px) {
          width: 480px;
        }
      `
    }
  }}
`

const ManipulateSection = styled.div`
  display: flex;

  > div {
    &:nth-child(1) {
      width: 75%;
    }

    &:nth-child(2) {
      width: 25%;
    }
  }
`
