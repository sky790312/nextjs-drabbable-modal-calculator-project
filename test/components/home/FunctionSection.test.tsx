import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { FunctionSection } from '@/components/home/FunctionSection'

describe('FunctionSection component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(withProviders(<FunctionSection />), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
