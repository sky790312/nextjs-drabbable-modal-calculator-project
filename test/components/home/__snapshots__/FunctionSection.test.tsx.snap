// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionSection component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz jOXVUR"
  >
    <div
      class="GlobalStyles__FlexCenterContainer-sc-1m5drgn-0 cmgQYV"
    >
      <button
        class="sc-bdfBwQ eCqVEX"
      >
        AC
      </button>
    </div>
    <div
      class="GlobalStyles__FlexCenterContainer-sc-1m5drgn-0 cmgQYV"
    >
      <button
        class="sc-bdfBwQ eCqVEX"
      >
        +/-
      </button>
    </div>
    <div
      class="GlobalStyles__FlexCenterContainer-sc-1m5drgn-0 cmgQYV"
    >
      <button
        class="sc-bdfBwQ eCqVEX"
      >
        %
      </button>
    </div>
  </div>
</DocumentFragment>
`;
